aws-distro-opentelemetry-node-autoinstrumentation/src/llo-handler.ts

```diff
@@ -0,0 +1,558 @@
+   * @returns
+   */
+  private emitLloAttributes(span: ReadableSpan, attributes: Attributes, eventTimestamp?: HrTime): void {
+    if (!attributes || Object.keys(attributes).length === 0) {
+      return;
+    }
+
+    //[] maybe add sanity check here from Python "has_llo_attrs"
+
+    const allMessages = this.collectAllLloMessages(span, attributes);
+    if (allMessages.length === 0) {
+      return;
+    }
+
+    // Group messages into input/output categories
+    const groupedMessages = this.groupMessagesByType(allMessages);
+
+    // Build event body
+    const eventBody: AnyValue = {};
+    if (groupedMessages.input.length > 0) {
+      eventBody.input = { messages: groupedMessages.input };
+    }
+    if (groupedMessages.output.length > 0) {
+      eventBody.output = { messages: groupedMessages.output };
+    }
+
+    if (Object.keys(eventBody).length === 0) {
+      return;
+    }
+
+    // Create and emit the event
+    const timestamp = eventTimestamp || span.endTime;
+    const eventLogger = this.eventLoggerProvider.getEventLogger(span.instrumentationLibrary.name);
+
+    // Hack - Workaround to add a custom-made Context to an Event so that the emitted event log
+    // has the correct associated traceId and spanId. This is needed because a ReadableSpan only
+    // provides its SpanContext, but does not provide access to its associated Context. We can only
+    // provide a Context to an Event, but not a SpanContext. Here we attach a custom Context that
+    // is associated to the ReadableSpan to mimic the ReadableSpan's actual Context.
+    //
+    // When a log record instance is created from this event, it will use the "custom Context" to
+    // extract the ReadableSpan from the "custom Context", then extract the SpanContext from the
+    // RedableSpan. This way, the emitted log event has the correct SpanContext (traceId, spanId).
+    // - https://github.com/open-telemetry/opentelemetry-js/blob/experimental/v0.57.1/experimental/packages/sdk-logs/src/LogRecord.ts#L101
+    // - https://github.com/open-telemetry/opentelemetry-js/blob/experimental/v0.57.1/api/src/trace/context-utils.ts#L78-L85
+    //
+    // We could omit the context field which is optional, but then the OTel EventLogger will assign
+    // the same `context.active()` and its associated SpanContext to each span from processSpans(),
+    // which would be incorrect since each span should have their own Context with unique SpanContext.
+    // - https://github.com/open-telemetry/opentelemetry-js/blob/experimental/v0.57.1/experimental/packages/sdk-events/src/EventLogger.ts#L34
+    const customContext = ROOT_CONTEXT.setValue(OTEL_SPAN_KEY, span);
+    const event: Event = {
+      name: span.instrumentationLibrary.name,
+      timestamp: timestamp,
+      data: eventBody,
+      context: customContext,
+    };
+
+    eventLogger.emit(event);
+    diag.debug('Emitted Gen AI Event with input/output message format');
```

Comment:
maybe remove this one seems not neccessary, will flood logs if turned on debugging mode but not really provide much value for troubleshooting